### PR TITLE
Shard snapshot API

### PR DESCRIPTION
**Pull request by ffuugoo**
Tracking issue: https://github.com/qdrant/qdrant/issues/2432

__TODO:__
- [x] Implement remote (URL-based) shard snapshot recovery
- [x] Copy shard state transition logic from [`_do_recover_from_snapshot`]
- [x] Add automatic shard snapshot recovery on shard snapshot upload (trivial, just braindead already to do it today)
- [x] Test ___everything___ 😵🔫
- [x] Implement integration tests ☠️🪦
- [x] _Always_ spawn a future for create/recover/upload APIs (e.g., see `helpers::time*` functions)? 🤔
- [x] Do not require shard to be _local_, when recovering shard snapshot 🤦‍♀️
  - I think `upload_shard_snapshot` would fail, if you'd try to recover a shard, that is not _local_ on the node
- [x] Return error `404`, if trying to restore a shard, that _does not exist in the replica set_
- [x] Move uploaded and remote-downloaded shard snapshots from the temporary/downloads directory to the shard snapshot directory
  - after _upload/download_, if shard is _local_ on the node
  - or after recovery, if local shard is created on the node during recovery
  - the logic for this is
    - if local shard exists, then user can list/retry/delete uploaded/downloaded snapshot using the API
    - and if local shard does not exist, then we remove failed snapshot automatically to prevent polluting temp/downloads directory
  - `std::fs::rename`/`tokio::fs::rename` only works if source and destination paths are on the same file system
    - [x] so `move` have to be implemented using `rename`/`copy`/`delete`
  - __UPD:__ This is currently done for shard upload, but temporarily reversed for remote shard download

[`_do_recover_from_snapshot`]: https://github.com/qdrant/qdrant/blob/master/lib/storage/src/content_manager/snapshots/recover.rs#L67-L288

__Further Improvements:__
- [ ] Test concurrent API calls and add write-locks if/where required
- [ ] Check for blocking calls in shard snapshot API implementation and replace them with async/`tokio::task::spawn_blocking`
- [ ] Unify and deduplicate `_do_recover_from_snapshot` logic with shard snapshot API
- [ ] Implement proper _pytest_ integration tests instead of `shard-snapshot-api.sh`
- [ ] Store basic collection config (e.g., ___vector size___) in shard snapshot and check them during recovery to prevent errors 🤔

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by ffuugoo** [comment]
@generall @timvisee Realized I need to report shard state to consensus and got stuck studying [`_do_recover_from_snapshot`] (which is a lot). Hope to wrap this up on Monday. 🤦‍♀️

Currently, recovering _local_ shard works (which overrides local shard). I think this should (might?) also work in distributed mode, but proper support for this is not implemented _yet_.

Also haven't yet finished recover from remote snapshot URL support. Current API works with snapshot files in local shard snapshot directory as `POST collections/<collection>/shards/<shard>/snapshots/<local snapshot filename>`

[`_do_recover_from_snapshot`]: https://github.com/qdrant/qdrant/blob/master/lib/storage/src/content_manager/snapshots/recover.rs#L67-L288

**Comment by ffuugoo** [comment]
@generall @timvisee Implemented remote file download and shard state transitions. What's left now is to test/debug everything and implement integration tests. And cleanup/refactor/unify all snapshotting code, cause there's some duplication between shard snapshot and full/collection snapshot code and also some bugs/improvements to full/collection snapshot code that I've spotted. But I'll probably do so in a separate PR, cause I'm already close to 1k limit (or was it 500 lines?) and I haven't even update OpenAPI spec yet.

**Comment by ffuugoo** [comment]
### TESTS

- **LIST**
  - list snapshots without any snapshots created
  - list snapshots with some snapshots created
  - try to list snapshots for a non existing collection
  - try to list snapshots for a non existing shard

- **DOWNLOAD**
  - download a snapshot
  - try to download a snapshot for a non existing collection
  - try to download a snapshot for a non existing shard
  - try to download a non existing snapshot

- **DELETE**
  - delete a snapshot
  - try to delete a snapshot for a non existing collection
  - try to delete a snapshot for a non existing shard
  - try to delete a non existing snapshot

- **CREATE**
  - create a snapshot
  - try to create a snapshot for a non existing collection
  - try to create a snapshot for a non existing shard
  - create multiple snapshots *concurrently*

- **RECOVER**
  - recover a snapshot
    - **single node and cluster**
    - local
    - remote
  - try to recover a snapshot for a non existing collection
    - local
    - remote
  - try to recover a snapshot for a non existing shard
    - local
    - remote
  - try to recover a non existing snapshot
    - local
    - remote
  - *concurrently* recover the same snapshot multiple times
    - **single node and cluster**
    - local
    - remote
  - *concurrently* recover multiple snapshots
    - **single node and cluster**
    - local
    - remote
    - local *and* remote
    - local *and* remote with the same filename
  - try to recover remote snapshot and cancel remote file download
    - ensure partially downloaded file is not preserved
  - try to recover remote snapshot with a weird URL/filename, e.g.:
    - with `/` or `:` character in the filename
    - or with a non UTF-8 filename

- **UPLOAD**
  - upload a snapshot
    - **single node and cluster**
  - try to upload a snapshot for a non existing collection
  - try to upload a snapshot for a non existing shard
  - try to upload a non existing snapshot
  - *concurrently* upload the same snapshot multiple times
    - **single node and cluster**
  - *concurrently* upload multiple snapshots
    - **single node and cluster**
  - try to upload a snapshot and cancel request while uploading a file
    - ensure partially downloaded file is not preserved
  - try to upload a snapshot and cancel request after uploading a file
  - try to upload a snapshot with a weird filename, e.g.:
    - with `/` or `:` character in the filename
    - or with a non UTF-8 filename

**Comment by ffuugoo** [comment]
@generall @timvisee Made a list of tests. Any ideas what to add? 🤔

File: src/actix/helpers.rs [hunk 1]
```diff
@@ -70,3 +72,132 @@ where
...
+
+impl From<CollectionError> for HttpError {
+    fn from(err: CollectionError) -> Self {
+        StorageError::from(err).into()
+    }
+}
+
+impl From<io::Error> for HttpError {
+    fn from(err: io::Error) -> Self {
+        StorageError::from(err).into() // TODO: Is this good enough?.. 🤔
```
Comment (timvisee) [inline src/actix/helpers.rs:201]: Yes :+1:

**Review by timvisee — APPROVED** [review]
Nice work!

No clear logic issues for now. Other than the TODOs still in the code this looks good!

I am somehow always getting a `null` creation time back though, for snapshots. It is not immediately clear to me why this is happening.

File: src/actix/api/snapshot_api.rs [hunk 2]
```diff
@@ -272,6 +279,221 @@ async fn delete_collection_snapshot(
...
+            .create_shard_snapshot(shard, &toc.temp_snapshots_path().join(SNAPSHOTS_TEMP_DIR))
+            .await?;
+
+        Ok(snapshot)
+    };
+
+    helpers::time_or_accept(future, query.wait.unwrap_or(true)).await
+}
+
+// TODO: `POST` or `PUT` (same as `recover_from_snapshot`)!?
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply src/actix/api/snapshot_api.rs:316]: Already fixed. Will push soon.

File: src/actix/api/snapshot_api.rs [hunk 3]
```diff
@@ -272,6 +279,221 @@ async fn delete_collection_snapshot(
...
+        )
+        .await?;
+
+        Ok(())
+    };
+
+    helpers::time_or_accept(future, query.wait.unwrap_or(true)).await
+}
+
+// TODO: `PUT` or `POST` (same as `upload_snapshot`)!?
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply src/actix/api/snapshot_api.rs:349]: Ditto.

**Comment by ffuugoo** [comment]
Implemented some [basic API tests](https://github.com/qdrant/qdrant/pull/2410/commits/26b010fa1dfe0ff8a494d67889796ce884a80cdc). Not perfect, but better than nothing. Really don't want to commit it without at least _some_ automated/reproducible tests.

@generall @timvisee

**Comment by timvisee** [comment]
> Not perfect, but better than nothing. Really don't want to merge this PR without at least _some_ automated/reproducible tests.

Yeah, good point.

> Implemented some [basic API tests](https://github.com/qdrant/qdrant/pull/2410/commits/26b010fa1dfe0ff8a494d67889796ce884a80cdc). Will finish them tomorrow morning.

Sounds good!

How about adding some tests [here](https://github.com/qdrant/qdrant/tree/master/tests/consensus_tests), I feel like it's easier to integrate than building our own shell script.

File: src/actix/helpers.rs [hunk 4]
```diff
@@ -70,3 +72,132 @@ where
         }
     }
 }
+
+pub async fn time<T, Fut>(future: Fut) -> impl actix_web::Responder
+where
+    Fut: Future<Output = HttpResult<T>>,
+    T: serde::Serialize,
+{
+    time_impl(async move { future.await.map(Some) }).await
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply src/actix/helpers.rs]: Fixed.

File: src/actix/api/snapshot_api.rs [hunk 5]
```diff
@@ -272,6 +280,263 @@ async fn delete_collection_snapshot(
...
+            .create_shard_snapshot(shard, &toc.temp_snapshots_path().join(SNAPSHOTS_TEMP_DIR))
+            .await?;
+
+        Ok(snapshot)
+    };
+
+    helpers::time_or_accept(future, query.wait.unwrap_or(true)).await
+}
+
+// TODO: `PUT` (same as `recover_from_snapshot`) or `POST`!?
```
Comment (timvisee) [inline src/actix/api/snapshot_api.rs:317]:
```suggestion
```

File: src/actix/api/snapshot_api.rs [hunk 6]
```diff
@@ -272,6 +280,263 @@ async fn delete_collection_snapshot(
...
+        )
+        .await?;
+
+        Ok(())
+    };
+
+    helpers::time_or_accept(future, query.wait.unwrap_or(true)).await
+}
+
+// TODO: `POST` (same as `upload_snapshot`) or `PUT`!?
```
Comment (timvisee) [inline src/actix/api/snapshot_api.rs:363]:
```suggestion
```

**Review by timvisee — APPROVED** [review]
Nice! Awesome to see you've covered so much tests already.

I think the only thing left to do, is to add it in our CI workflow for automated testing.

File: src/actix/api/snapshot_api.rs (hunk 5 above)
Comment (ffuugoo) [reply src/actix/api/snapshot_api.rs:317]: Kinda left it on purpose, to consider it later once again.

File: src/actix/helpers.rs (hunk 1 above)
Comment (ffuugoo) [reply src/actix/helpers.rs:202]: Left the `TODO` for further consideration.

File: src/actix/api/snapshot_api.rs (hunk 5 above)
Comment (timvisee) [reply src/actix/api/snapshot_api.rs:317]: :+1: 

Maybe we can deprecate it eventually, and remove it all together in a major release (if we ever do).

**Comment by ffuugoo** [comment]
Noticed a bunch of small issues with `recover_shard_snapshot` and `upload_shard_snapshot`. Marked them all under TODO in the PR description. Plan to fix them now.

**Comment by ffuugoo** [comment]
> Noticed a bunch of small issues with `recover_shard_snapshot` and `upload_shard_snapshot`. Marked them all under TODO in the PR description. Plan to fix them now.

Hopefully, these are all fixed now. 🤞

File: lib/collection/src/common/file_utils.rs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+use fs_extra::dir::CopyOptions;
+
+use crate::operations::types::{CollectionError, CollectionResult};
+
+/// Move directory from one location to another.
+/// Handles the case when the source and destination are on different filesystems.
+pub async fn move_dir(from: impl Into<PathBuf>, to: impl Into<PathBuf>) -> CollectionResult<()> {
+    // Try to rename first and fallback to copy to prevert TOCTOU
+    let from = from.into();
+    let to = to.into();
```
Comment (ffuugoo) [inline lib/collection/src/common/file_utils.rs:12]:
```suggestion
pub async fn move_dir(from: impl AsRef<Path>, to: impl AsRef<Path>) -> CollectionResult<()> {
    // Try to rename first and fallback to copy to prevert TOCTOU
    let from = from.as_ref();
    let to = to.as_ref();
```

Comment (ffuugoo) [reply lib/collection/src/common/file_utils.rs:12]: Disregard this. There's `spawn_blocking`, which requires owned data.

**Review by generall — APPROVED** [review]

**Comment by ffuugoo** [comment]
@generall I'll merge this once tests pass.

**Pull request merged**